server/socketManager.js

```diff
@@ -0,0 +1,373 @@
+   */
+  init() {
+    this.on('request', req => {
+      if (req.method === 'POST' && req.pathname === '/') return;
+
+      this.logger.debug(
+        'Request for method=%s path=%s (%s).',
+        req.method,
+        req.pathname,
+        req.socket.remoteAddress
+      );
+    });
+
+    this.on('listening', address => {
+      this.logger.info(
+        'SocketManager server listening on %s (port=%d).',
+        address.address,
+        address.port
+      );
+    });
+  }
+
+  /*
+   * handle new socket clients that connect to server
+   * @param {WebSocket} socket
+   * @returns {null}
+   */
+  handleSocket(socket) {
+    socket.hook('auth', (...args) => {
+      if (socket.channel('auth')) throw new Error('Already authed.');
+
+      if (!this.options.noAuth) {
+        const valid = new Validator(args);
+        const key = valid.str(0, '');
+        if (key.length > 255) throw new Error('Invalid API key.');
+
+        const data = Buffer.from(key, 'utf8');
+        const hash = sha256.digest(data);
+
+        if (!ccmp(hash, this.options.apiHash))
+          throw new Error('Invalid API key.');
+      }
+
+      socket.join('auth');
+      this.logger.info('Successful auth from %s.', socket.host);
+      this.handleAuth(socket);
+      return null;
+    });
+  }
+
+  /*
+   * Utility to verify that the pathname
+   * matches with the id of a client on the class
+   * @private
+   * @param {string} pathname - pathname to validate
+   * @returns {string} id
+   */
+  getIdFromPath(pathname) {
+    assert(
+      pathname[0] === '/' && pathname[pathname.length - 1] === '/',
+      'Invalid pathname'
+    );
+    const id = pathname.slice(1, pathname.length - 1);
+    if (!this.clients.has(id))
+      this.logger.warning(
+        'getIdFromPath: id %s from pathname %s does not exist',
+        id,
+        pathname
+      );
+
+    return id;
+  }
+
+  /*
+   * Handle new auth'd websocket.
+   * @private
+   * @param {WebSocket} socket
+   */
+  handleAuth(socket) {
+    this.logger.info('Preparing socket from %s', socket.url);
+    const parseEvent = event => {
+      assert(typeof event === 'string', 'Must pass a string to getClient');
+      // defaulting to node client if there is no client prefix
+      let parsedEvent = event;
+      let id = 'node';
+
+      // check if first full word is a valid type
+      const split = event.indexOf(' ');
+      const prefix = event.slice(0, split);
+      if (this.types.indexOf(prefix) !== -1) {
+        id = prefix;
+        parsedEvent = event.slice(split + 1);
+      }
+      return [id, parsedEvent];
+    };
+
+    // creating a unique channel name for subscriptions
+    // each channel needs to be unique to the clientId
+    // (one of accepted types, e.g. wallet or node),
+    // the id of socket (i.e. from the socket's path)
+    // event being subscribed to and the event fired when event
+    // is heard.
+    const getChannelName = (clientId, id, event, responseEvent) => {
```

Comment:
I think that this function should be moved out from being defined within another function so that it can be used anywhere, including in the tests

Reply:
yep, agreed. Moved it out already in the next commit 👍